galsim/zernike.py

```diff
@@ -239,17 +433,65 @@ class Zernike(object):
     @param R_inner  Inner radius.  [default: 0.0]
     """
     def __init__(self, coef, R_outer=1.0, R_inner=0.0):
-        self.coef = np.array(coef)
+        self.coef = np.asarray(coef)
         self.R_outer = float(R_outer)
         self.R_inner = float(R_inner)
 
-        self._coef_array = np.dot(
-                _noll_coef_array(len(self.coef)-1, R_inner/R_outer),
-                self.coef[1:])
+    # _coef_array property only exists to support the deprecated OpticalPSF.coef_array attribute.
+    # It can be deleted in version 2.0.
+    @lazy_property
+    def _coef_array(self):
+        _coef_array = np.dot(
+            _noll_coef_array(len(self.coef)-1, self.R_inner/self.R_outer),
+            self.coef[1:]
+        )
 
-        if R_outer != 1.0:
-            shape = self._coef_array.shape
-            self._coef_array /= R_outer**np.sum(np.mgrid[0:2*shape[0]:2, 0:shape[1]], axis=0)
+        if self.R_outer != 1.0:
+            shape = _coef_array.shape
+            _coef_array /= self.R_outer**np.sum(np.mgrid[0:2*shape[0]:2, 0:shape[1]], axis=0)
+        return _coef_array
+
+    @lazy_property
+    def _coef_array_xy(self):
+        _coef_array_xy = np.dot(
+            _noll_coef_array_xy(len(self.coef)-1, self.R_inner/self.R_outer),
+            self.coef[1:]
+        )
+
+        if self.R_outer != 1.0:
+            shape = _coef_array_xy.shape
+            _coef_array_xy /= self.R_outer**(np.sum(np.mgrid[0:shape[0], 0:shape[1]], axis=0))
+        return _coef_array_xy
+
+    @lazy_property
+    def _gradX(self):
+        j = len(self.coef)-1
+        newCoef = np.hstack([
+            [0],
+            np.dot(
+                _noll_coef_array_gradx(j, self.R_inner/self.R_outer),
+                self.coef[1:]
+            )
+        ])
+        # Handle R_outer with
+        # df/dx = df/d(x/R) * d(x/R)/dx = df/d(x/R) * 1/R
+        newCoef /= self.R_outer
+        return Zernike(newCoef, R_outer=self.R_outer, R_inner=self.R_inner)
+
+    @lazy_property
+    def _gradY(self):
+        j = len(self.coef)-1
+        newCoef = np.hstack([
+            [0],
+            np.dot(
+                _noll_coef_array_grady(j, self.R_inner/self.R_outer),
+                self.coef[1:]
+            )
```

Comment:
I see a few of these `np.dot` calls.  I always found it clearer to write `A.dot(B)` rather than `np.dot(A,B)`.  Infix rather than prefix is I guess closer to the infix math notation `A * B`.  If you prefer to leave it this way, it's fine, but I just thought I'd put that out there.

Reply:
I guess I normally prefer `np.dot(A, B)` for its symmetry.  That said, in this case, since the dots are really matrix-vector multiplication, the operation isn't symmetric, so I went ahead and switched 'em.